emscripten.py

```diff
@@ -2400,9 +2400,11 @@ def create_asm_consts_wasm(forwarded_json, metadata):
 
   asm_const_funcs = []
 
-  if all_sigs:
-    # emit the signature-reading helper function only if we have any EM_ASM
-    # functions in the module
+  # emit the signature-reading helper function only if we have any EM_ASM
+  # functions in the module. we also need to emit it for pthreads anyhow as
+  # the proxying code there, which is always emitted, supports EM_ASMs and
+  # requires this support code
```

Comment:
But if the application doesn't have any EM_ASM usage at all why would the proxy code ever actually use this?  Does it use for some other purpose?

Reply:
It would not be used in practice, and it would be good to not emit it. However, it's not trivial to get the information that there are no EM_ASMs into the JS compiler, so I didn't pursue that.

Reply:
Sorry I still don't understand.   Why would not including this code here cause a runtime error?  We know here that this code won't be needed right?   So any proxying code that depends in this stuff will not fail at runtime right?  Or is closure checking that there are no undefined functions? 

Another way of putting it:  If you don't make this change what is the error that results?

Reply:
There is nothing that requires this function to  be a special dynamically emitted function from emscripten.py? How about moving all of this from being dynamically generated into library_pthread.js, and doing

```diff
diff --git a/src/library_pthread.js b/src/library_pthread.js
index 07a54d6a3..cf579e661 100644
--- a/src/library_pthread.js
+++ b/src/library_pthread.js
@@ -1295,7 +1295,7 @@ var LibraryPThread = {
 
   emscripten_receive_on_main_thread_js_callArgs: '=[]',
 
-  emscripten_receive_on_main_thread_js__deps: ['emscripten_proxy_to_main_thread_js', 'emscripten_receive_on_main_thread_js_callArgs'],
+  emscripten_receive_on_main_thread_js__deps: ['emscripten_proxy_to_main_thread_js', 'emscripten_receive_on_main_thread_js_callArgs', '$readAsmConstArgs'],
   emscripten_receive_on_main_thread_js: function(index, numCallArgs, args) {
     _emscripten_receive_on_main_thread_js_callArgs.length = numCallArgs;
     var b = args >> 3;
````
That seems like a fully linking aware solution? Then if there are any EM_ASMs, force-include `$readAsmConstArgs` like is done above for `_proxy_main`?

Reply:
> Or is closure checking that there are no undefined functions?

Yeah, closure does not allow any undefined references in the output code.